HPCC-20223 Remove legacy token from WU/DFUWU, and write new signed token

RussWhitehead opened the pull request:
Removes support for the old legacy workunit token
Adds methods to create, identify, validate, and extract from new workunit distributed access token
Creates the new DAToken when workunits are created
Allows the DAToken to be unsigned if no certificates configured
ESP Services set the DAToken as a password in the ISecUser
ESP Verifies DAToken if provided as an ISecUser password
Dali allows getPermissions requests even if no password provided, authenticates if it is

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

RussWhitehead committed:
HPCC-20223 Remove legacy token from WU/DFUWU, and write new signed token

Removes support for the old legacy workunit token
Adds methods to create, identify, validate, and extract from new workunit distributed access token
Creates the new DAToken when workunits are created
Allows the DAToken to be unsigned if no certificates configured
ESP Services set the DAToken as a password in the ISecUser
ESP Verifies DAToken if provided as an ISecUser password
Dali allows getPermissions requests even if no password provided, authenticates if it is

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@afishbeck Please rereview. Note I successfully ran the regression suite with and with certificates enabled.
@richardkchapman @jakesmith @ghalliday

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9bc81181f7b14fe72284672ba809927b72fa79a3
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |1 |0 |0 |0 |1 |
|wutoolTest(Cassandra) |1 |0 |0 |0 |1 |




Wutool error(s):

Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |0 |11 |
|setup (thor) |11 |0 |11 |
|setup (roxie) |11 |0 |11 |
|test (hthor) |805 |14 |791 |
|test (thor) |729 |11 |718 |
|test (roxie) |878 |14 |864 |


[Errors:](https://gist.github.com/755c07e1faf8b9b3b4e4357f01c01179)
- setup
  - hthor
    - 1\. [roxie_echo](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_echo-no-wuid-log), 2\. [roxie_keepwhitespace](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_keepwhitespace-no-wuid-log), 3\. [setup](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setup-no-wuid-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setup_fetch-no-wuid-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupdict-no-wuid-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setuppersist-no-wuid-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupsearchindex-no-wuid-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupsq-no-wuid-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupwordindex-no-wuid-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupxml-no-wuid-log), 11\. [soapbase](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-soapbase-no-wuid-log), 
  - thor
    - 1\. [roxie_echo](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_echo-no-wuid-log), 2\. [roxie_keepwhitespace](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_keepwhitespace-no-wuid-log), 3\. [setup](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setup-no-wuid-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setup_fetch-no-wuid-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupdict-no-wuid-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setuppersist-no-wuid-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupsearchindex-no-wuid-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupsq-no-wuid-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupwordindex-no-wuid-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-setupxml-no-wuid-log), 11\. [soapbase](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-soapbase-no-wuid-log), 
  - roxie
    - 1\. [roxie_echo](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_echo-no-wuid-log), 2\. [roxie_keepwhitespace](https://gist.github.com/HPCCSmoketest/755c07e1faf8b9b3b4e4357f01c01179#file-roxie_keepwh
 ... ( comment is too long, 56 lines, 332405 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  183 sec (00:03:03) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 2432 sec (00:40:32) |   15 sec (00:00:15) | 2800 sec (00:46:40) |

```

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d035c7c60870a2fbb795d39c0b243571b2095154
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |18 |1 |0 |0 |
|wutoolTest(Cassandra) |19 |18 |1 |0 |0 |




Wutool error(s):
Dali:1) test: WuTool::testCopy (F) line: 1279 /mnt/disk1/home/vamosax/smoketest/PR-11628/HPCC-Platform/tools/wutool/wutool.cpp

Dali:assertion failed

Cassandra:1) test: WuTool::testCopy (F) line: 1279 /mnt/disk1/home/vamosax/smoketest/PR-11628/HPCC-Platform/tools/wutool/wutool.cpp

Cassandra:assertion failed


Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  185 sec (00:03:05) |   93 sec (00:01:33) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1349 sec (00:22:29) |   20 sec (00:00:20) | 1717 sec (00:28:37) |

RussWhitehead commented:
Fixed ESP core that occurred when running without authentication (LDAP) enabled.

RussWhitehead commented:
All tests pass. Regression suite and wutoolTest with no authentication, LDAP without certs, and LDAP with certs. 
@afishbeck Please rereview
@richardkchapman @ghalliday @jakesmith @stuartort

afishbeck commented on common/workunit/workunit.cpp:
```diff
@@ -6329,17 +6331,47 @@ void CLocalWorkUnit::setDebugAgentListenerIP(const char * ip)
+            pFN = pDSM->queryPrivateKeyFile();
+        else
+            ++pFN;
+        datoken.append(pFN).append(';');//include private key file name used to encrypt
```
Is there a JIRA open about this identifying which public key to use to verify, rather than the name of the private key being used?  So that if we want to we can allow the public keys to live longer?

afishbeck replied:
@jakesmith @RussWhitehead Yes, the concept looks good.  I wasn't sure it would make it into 7.0 but its great that it is, since its a major bit of configuration for these features.

I would think that the keypairName could be configured for signing workunits, and used as the keyname part of the token.  The names could then be versioned if we wanted to keep them around for verification even after they are no longer used for signing.

afishbeck commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+        if (finger && *finger)
+        {
+            StringBuffer sig(++finger);
+            if (!pDSM->digiVerify(token, sig))
```
Maybe the same as above, but is there a JIRA open about digiVerify being told which public key to use for this particular signatue?

afishbeck reviewed (commented):
@looks good, but still wonder about whether we are going to use the key name to identify the verification key.  Is there already a separate JIRA for that?

afishbeck commented:
@RussWhitehead

RussWhitehead commented:
@afishbeck We should use https://track.hpccsystems.com/browse/HPCC-20158 to address the certificate stuff
@richardkchapman Please merge

richardkchapman commented:
@ghalliday Please sanity check

ghalliday commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -667,22 +686,41 @@ bool CLdapSecManager::authenticate(ISecUser* user)
     }
 
     //Verify provided signature if present
```
Please can you add a comment clarifying how the signature differs from the password, and when each are used.

ghalliday commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -667,22 +686,41 @@ bool CLdapSecManager::authenticate(ISecUser* user)
-            user->setAuthenticateStatus(AS_INVALID_CREDENTIALS);
-            WARNLOG("Invalid digital signature for user %s", user->getName());
-            return false;
+            if (streq(cachedUser->credentials().getSignature(), user->credentials().getSignature()))
```
could signature have been null last time when it was cached?

RussWhitehead replied:
I see no obvious code path where this is possible and did not come across it in my testing

ghalliday commented on system/security/cryptohelper/digisign.hpp:
```diff
@@ -37,6 +37,9 @@ interface IDigitalSignatureManager : extends IInterface //Public/Private key mes
 public:
     virtual bool isDigiSignerConfigured() = 0;
     virtual bool isDigiVerifierConfigured() = 0;
+    virtual void setKeyFileNames(const char * pub, const char * priv) = 0;
```
This shouldn't be in the public interface.  It can be a member of the class that is created.

RussWhitehead replied:
agreed

ghalliday commented on tools/swapnode/swapnodelib.cpp:
```diff
@@ -84,16 +84,15 @@ bool WuResubmit(const char *wuid)
-    extractToken(token.str(), wuid, user, password);
+    SCMStringBuffer daToken;
+    StringBuffer user;
+    wu->getWorkunitDistributedAccessToken(daToken);//TODO CAn't i just use wu->queryUser()
```
typo in comment

RussWhitehead replied:
fixed

ghalliday reviewed (commented):
@RussWhitehead a reasonable number of fairly minor comments, and a couple more significant (return type of function and dependency on workunit.so)

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6329,17 +6331,47 @@ void CLocalWorkUnit::setDebugAgentListenerIP(const char * ip)
 }
 
-IStringVal& CLocalWorkUnit::getSecurityToken(IStringVal &str) const
+void CLocalWorkUnit::getWorkunitDistributedAccessToken(IStringVal & datoken) const
```
agreed

RussWhitehead replied:
agreed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+    }
+
+    //Extract the string between [ and ]
+    const char * pEndBracket = strchr(distributedAccessToken + 5, ']');
```
agreed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+    if (privKey)
+    {
+        const char * finger = pEndBracket;
+        if (finger)
```
agreed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+        if (finger)
+        {
+            ++finger;
+            while (*finger && *finger != ';')
```
This code is copying everything up to the ';' so using strchr doesn't really help. Prefer to leave it as is

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+CriticalSection verifyWorkunitDATokenCrit;
+int verifyWorkunitDAToken(const char * distributedAccessToken)
+{
+    CriticalBlock block(verifyWorkunitDATokenCrit);
```
Unneeded relic from the past, removed the critical section

RussWhitehead commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -650,6 +650,25 @@ bool CLdapSecManager::authenticate(ISecUser* user)
         return false;
     }
 
+    //Check if the password is a "real" password, or a workunit distributed access token
```
This is a private method called from several places within this same CLdapSecManager

ghalliday replied:
I will try and understand the structure a bit more, but it doesn't feel right.

RussWhitehead commented:
@ghalliday I implemented your suggestions, and added code to dasess.cpp to not serialize the user password to Dali getPermissions if that password is a workunit token. Retested to make sure nothing broke
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9fff83d723b14966b2fd0e7eded305b9ed4fe7a2
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  185 sec (00:03:05) |   93 sec (00:01:33) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1339 sec (00:22:19) |   20 sec (00:00:20) | 1725 sec (00:28:45) |

ghalliday commented on dali/base/CMakeLists.txt:
```diff
@@ -58,6 +58,7 @@ set (    INCLUDES
 include_directories ( 
          ./../../common/deftype 
          ./../../common/remote 
+         ./../../common/workunit
```
NO! Dali should not be dependent on workunit.

RussWhitehead replied:
Dali does not link with workunit, it only includes that file so it can resolve isWorkunitDAToken()

ghalliday replied:
It still shows an incorrect dependency.

ghalliday reviewed (commented):
@RussWhitehead a few quick comments in response.

ghalliday commented:
This will also need corresponding changes to the LN repo.  Currently eclserver fails to compile (Pending.cpp)

richardkchapman commented on common/workunit/workunit.hpp:
```diff
@@ -1597,15 +1596,28 @@ extern WORKUNIT_API IWUResult * updateWorkUnitResult(IWorkUnit * w, const char *
+//workunit distributed access token support
+extern WORKUNIT_API int verifyWorkunitDAToken(const char * distributedAccessToken);
+extern WORKUNIT_API bool extractFromWorkunitDAToken(const char * token, StringBuffer * wuid, StringBuffer * user, StringBuffer * privKey);
+inline bool isWorkunitDAToken(const char * distributedAccessToken)
```
It's inline to avoid a cyclic dependency between workunit and dalibase

RussWhitehead replied:
Richard is correct

ghalliday replied:
That dependency should not exist - but I'll address that elsewhere.

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6838,117 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+        const char * finger = start;
+        ++finger;
+        while (*finger && *finger != ';')
+            privKey->append(1, (const char *)finger++);
```
removed cast

RussWhitehead commented on system/security/cryptohelper/digisign.cpp:
```diff
@@ -367,7 +367,7 @@ extern "C"
         IDigitalSignatureManager * pDSM = createDigitalSignatureManagerInstanceFromKeys(publicKeyBuff, privateKeyBuff, _passPhrase);
         if (pDSM)
-            pDSM->setKeyFileNames(_pubKey, _privKey);
+            dynamic_cast<CDigitalSignatureManager*>(pDSM)->setKeyFileNames(_pubKey, _privKey);
```
Agreed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+//and signed by the given wuid and user, and that the workunit is still active
+//Returns:
+//   0 : Success, token is valid and workunit is active
+//   1 : Signature does not verify (wuid/username don't match, or signature doesnt verify)
```
fixed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+    if (!extractFromWorkunitDAToken(distributedAccessToken, &tokWuid, &tokUser, nullptr))//get the wuid and user
+    {
+        //Not a valid workunit distributed access token
+        return false;
```
Good catch, converted to enum

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+            token.append(1, finger);
+        token.append(1, finger);//append ;
+
+        if (finger && *finger)
```
fixed

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+    }
+
+    //Verify user matches
+    if (!streq(cw->queryUser(), tokUser.str()))
```
Fixed

RussWhitehead commented on system/security/LdapSecurity/CMakeLists.txt:
```diff
@@ -57,6 +58,7 @@ target_link_libraries ( LdapSecurity
          jlib
          cryptohelper
          dalibase
+         workunit
```
It needs to verify the workunit token, which requires checking if the workunit is active

RussWhitehead commented:
@ghalliday Please rereview

ghalliday commented on system/security/cryptohelper/digisign.cpp:
```diff
@@ -122,34 +122,34 @@ class CDigitalSignatureManager : implements IDigitalSignatureManager, public CIn
     }
 
-    void setKeyFileNames(const char * pub, const char * priv)
+    virtual void setKeyFileNames(const char * pub, const char * priv)
```
no need for this to be virtual

ghalliday commented on system/security/cryptohelper/digisign.cpp:
```diff
@@ -236,6 +254,20 @@ class CDigitalSignatureManager : implements IDigitalSignatureManager, public CIn
         return false;
     }
 
+    void setKeyFileNames(const char * pub, const char * priv)
```
That is only ever called from code that is conditional on the other class being created.
You could have a common base class between the two, and structure the code differently, but as it stands it isn't used.

ghalliday reviewed (commented):
I have traced through the code some more and these are the conclusions I have come to:

1. Dali should never see a hpcc[] token.  If the request comes in from outside the system (which should really should be locked down anyway), then it should have a username/password.  If it comes in from esp/and engine it should have a signature.

2. Ldap shouldn't be authenticating users when it checks permissions.  If ldap itself doesn't require it, neither should we (since the user could go straight to ldap to do the same checks).

3. The only place I could find that needed to check for the hpcc[] style token was EspHttpBinding::authenticate().  That is where the code that recognises it should live, and move it out of ldapsecurity.  All other instances I could find either fit under (1) or (2).

@afishbeck @RussWhitehead what do you think?

ghalliday commented:
NOTE: This also needs to be linked to a PR on the LN repository.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a339638a5b53531841175ae20f6ff365081a6d06
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  185 sec (00:03:05) |   93 sec (00:01:33) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1347 sec (00:22:27) |   20 sec (00:00:20) | 1724 sec (00:28:44) |

RussWhitehead commented:
@ghalliday Please re-review, I believe this addresses all of your comments

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5495b06a7d9a286f0c5a48a59c8022aa73826463
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   11 sec (00:00:11) |  185 sec (00:03:05) |   94 sec (00:01:34) |   19 sec (00:00:19) |   40 sec (00:00:40) | 1329 sec (00:22:09) |   18 sec (00:00:18) | 1696 sec (00:28:16) |

RussWhitehead commented:
@ghalliday This includes your recommendations from a few days ago, and a rebase onto the latest master
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a2542b7cb92b7c818d604d892d1dd245a113b024
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  191 sec (00:03:11) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1361 sec (00:22:41) |   19 sec (00:00:19) | 1752 sec (00:29:12) |

ghalliday commented on system/security/cryptohelper/digisign.cpp:
```diff
@@ -136,6 +138,22 @@ class CDigitalSignatureManager : public CSimpleInterfaceOf<IDigitalSignatureMana
         return digiVerify(b64Signature, strlen(text), text);
     }
+
+    void setKeyFileNames(const char * pub, const char * priv)
```
With Jake's refactoring this should be obtainable from the pubKey/privKey class, so no need to set up like this.

ghalliday replied:
I think his changes still mean you can remove this call.  E.g. You can implement the function as:

    virtual const char * queryPublicKeyFile() const override
    {
        return pubKey->queryKeyName();
    }

ghalliday reviewed (commented):
@RussWhitehead please see review comments - mainly caused by Jake's factoring of the code.

I am still not sure about which components the hpcc[] style token should be accepted and verified.  That will become more obvious as this change gets tested.  I may be proved wrong that dali needs to accept and verify the tokens, but I would prefer it structured differently if it is.

I am also not quite sure I follow the role of signatures and caching, but that is not directly related to this PR.

Once the issues from the rebase are resolved I think this is a ready for merging and a good starting point for seeing what problems we can find with the system.

RussWhitehead commented:
@ghalliday I addressed most of your issues. I am not sure about the usefulness of the user signature any more, and will address with https://track.hpccsystems.com/browse/HPCC-20041 . I don't think Dali needs to accept the HPCC[] token, since the only interface accepting credentials is getPermissions, and it uses the Dali Token (scope;user;timestamp) to validate the caller. Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fca3082bd51db124c718721c496ecff44478cfc6
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  186 sec (00:03:06) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1338 sec (00:22:18) |   18 sec (00:00:18) | 1711 sec (00:28:31) |

RussWhitehead commented:
@ghalliday Also fixed an exception thrown when the passphrase is not decrypted, and a "return new" that was inadvertently left in before my temp code to save the key file names. Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 65e89da18c4285f52686df119812df33601730aa
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  186 sec (00:03:06) |   95 sec (00:01:35) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1534 sec (00:25:34) |   19 sec (00:00:19) | 1909 sec (00:31:49) |

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -6329,17 +6331,48 @@ void CLocalWorkUnit::setDebugAgentListenerIP(const char * ip)
+        datoken.append(pFN).append(';');//include private key file name used to encrypt
+
+        StringBuffer b64Signature;
+        if (pDSM->digiSign(datoken, b64Signature))
```
@RussWhitehead the parameters need to be reversed.

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -6805,6 +6838,119 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+    //Verify user matches
+    if (!isEmptyString(cw->queryUser()) || !isEmptyString(tokUser.str()))
+    {
+        if (!streq(cw->queryUser(), tokUser.str()))
```
need to be careful - cw->queryUser() could be null - although someone would need to be deliberately trying to break the system.  E.g. create a workunit on a system with no security, and then send in a request using a token that does include a user.

ghalliday commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -701,6 +702,27 @@ bool EspHttpBinding::basicAuth(IEspContext* ctx)
+        if (state == wuTokenValid)
+        {
+            user->setAuthenticateStatus(AS_AUTHENTICATED);
+        }
```
Should this also clear the password?  Where is the password cleared if the user is authenticated?

ghalliday reviewed (commented):
@RussWhitehead a few final comments

ghalliday commented:
Looks like there is also a clash with a fix fro some windows build issues.

richardkchapman commented:
May clash with Jake's fix in https://github.com/hpcc-systems/HPCC-Platform/pull/11687

RussWhitehead commented:
@ghalliday Rebased and squashed. Please merge along with https://github.com/hpcc-systems/LN/pull/888

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 12f3cb5ffe06a6df572715000e84defb568aa99b
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  184 sec (00:03:04) |   95 sec (00:01:35) |   19 sec (00:00:19) |   42 sec (00:00:42) | 1359 sec (00:22:39) |   19 sec (00:00:19) | 1734 sec (00:28:54) |
